libminifi/include/SchedulingAgent.h

```diff
@@ -190,11 +207,27 @@ class SchedulingAgent {
   std::shared_ptr<core::controller::ControllerServiceProvider> controller_service_provider_;
 
  private:
+  struct SchedulingInfo {
+    std::chrono::time_point<std::chrono::steady_clock> start_time_ = std::chrono::steady_clock::now();
+    mutable std::chrono::time_point<std::chrono::steady_clock> last_alert_time_ = std::chrono::steady_clock::now();
+    std::string name_;
+    std::string uuid_;
+
+    explicit SchedulingInfo(const std::shared_ptr<core::Processor> &processor) :
+      name_(processor->getName()),
+      uuid_(processor->getUUIDStr()) {}
+
+    bool operator <(const SchedulingInfo& o) const {
+      return std::tie(start_time_, name_, uuid_) < std::tie(o.start_time_, o.name_, o.uuid_);
+    }
+  };
+
   // Logger
   std::shared_ptr<logging::Logger> logger_;
-  // Prevent default copy constructor and assignment operation
-  // Only support pass by reference or pointer
-
+  mutable std::mutex watchdog_mtx_;  // used to protect the set below
+  std::set<SchedulingInfo> scheduled_processors_;  // set was chosen to avoid iterator invalidation
+  std::unique_ptr<utils::CallBackTimer> watchDogTimer_;
+  int64_t alert_time_;  // msec
```

Comment:
unit in the type > unit in the name > unit in comment > no indication of unit in code
I suggest changing the type to `std::chrono::milliseconds` and only converting to integral when we're passing the value through an incompatible API. (Like the logging in `watchDogFunc`)

Reply:
Done